[Bug] undici 8 broke my calendar

### Environment


####  System Information  ####
- SYSTEM:   manufacturer: Apple Inc.; model: MacBookPro18,3; virtual: false; MM: v2.36.0-develop
- OS:       platform: darwin; distro: macOS; release: 26.4; arch: arm64; kernel: 25.4.0
- VERSIONS: electron: 41.1.1; used node: 22.22.2; installed node: 22.22.2; npm: 10.9.7; pm2: 6.0.14
- ENV:      XDG_SESSION_TYPE: undefined; MM_CONFIG_FILE: undefined
            WAYLAND_DISPLAY:  wayland-1; DISPLAY: undefined; ELECTRON_ENABLE_GPU: undefined
- RAM:      total: 32768.00 MB; free: 785.58 MB; used: 31982.42 MB
- OTHERS:   uptime: 182 minutes; timeZone: Europe/Berlin


### Which start option are you using?

node --run start

### Are you using PM2?

No

### Module

None

### Have you tried disabling other modules?

- [x] Yes
- [ ] No

### Have you searched if someone else has already reported the issue on the forum or in the issues?

- [x] Yes

### What did you do?

updated my mirror after the latest release and got only one of my calendars to work... 

### What did you expect to happen?

calendar loads, just like the other

### What actually happened?


[2026-04-05 21:56:38.958] [WARN]  [http_fetcher] https://xxxxxxxx.de/caldav/kalender/prwvrm/ - Network error: fetch failed Retry #1 in 15s.
[2026-04-05 21:56:38.959] [ERROR] [calendar] Calendar Error. Could not fetch calendar:  https://xxxxxxx.de/caldav/kalender/prwvrm/ Network error: fetch failed


### Additional comments

strangely my other calendar from the same url works like a charm.... will look into this maybe, for now i downgrade undici to v7 on my setup

### Participation

- [ ] I am willing to submit a pull request for this change.

--- a/js/http_fetcher.js
+++ b/js/http_fetcher.js
@@ -1,5 +1,5 @@
 const { EventEmitter } = require("node:events");
-const { Agent } = require("undici");
+const { fetch: undiciFetch, Agent } = require("undici");
 const Log = require("logger");
 const { getUserAgent } = require("#server_functions");
 
@@ -263,8 +263,13 @@ class HTTPFetcher extends EventEmitter {
 		const timeoutId = setTimeout(() => controller.abort(), this.timeout);
 
 		try {
-			const response = await fetch(this.url, {
-				...this.getRequestOptions(),
+			const requestOptions = this.getRequestOptions();
+			// Use undici.fetch when a custom dispatcher is present (e.g. selfSignedCert),
+			// because Node's global fetch and npm undici@8 Agents are incompatible.
+			// For regular requests, use globalThis.fetch so MSW and other interceptors work.
+			const fetchFn = requestOptions.dispatcher ? undiciFetch : globalThis.fetch;
+			const response = await fetchFn(this.url, {
+				...requestOptions,
 				signal: controller.signal
 			});
 

--- a/tests/unit/functions/http_fetcher_spec.js
+++ b/tests/unit/functions/http_fetcher_spec.js
@@ -440,3 +440,32 @@ describe("fetch() method", () => {
 		expect(errorInfo.errorType).toBe("NETWORK_ERROR");
 	});
 });
+
+describe("selfSignedCert dispatcher", () => {
+	const { Agent } = require("undici");
+
+	it("should set rejectUnauthorized=false when selfSignedCert is true", () => {
+		fetcher = new HTTPFetcher(TEST_URL, {
+			reloadInterval: 60000,
+			selfSignedCert: true
+		});
+
+		const options = fetcher.getRequestOptions();
+
+		expect(options.dispatcher).toBeInstanceOf(Agent);
+		const agentOptionsSymbol = Object.getOwnPropertySymbols(options.dispatcher).find((s) => s.description === "options");
+		const dispatcherOptions = options.dispatcher[agentOptionsSymbol];
+		expect(dispatcherOptions.connect.rejectUnauthorized).toBe(false);
+	});
+
+	it("should not set a dispatcher when selfSignedCert is false", () => {
+		fetcher = new HTTPFetcher(TEST_URL, {
+			reloadInterval: 60000,
+			selfSignedCert: false
+		});
+
+		const options = fetcher.getRequestOptions();
+
+		expect(options.dispatcher).toBeUndefined();
+	});
+});

